### PR TITLE
feat(core): Implement Dynamic Parameters within regular nodes used as AI Tools

### PR DESCRIPTION
**Pull request by OlegIvaniv**
## Summary

This PR introduces a new feature that allows LLM to dynamically interact with n8n nodes when used as tools. The core of this feature is the new `$fromAI()` function, which allows for the definition of parameters that can be populated by AI models during workflow execution.

https://github.com/user-attachments/assets/60d293ec-c126-4fa6-bfcd-8ba9acfefad4

## Key Changes

1. Introduced a new `$fromAI()` function for defining AI-powered parameters in node configurations.
2. Updated the `WorkflowDataProxy` to handle `$fromAI()` calls and retrieve values passed by LLMs.
3. Modified the `createNodeAsTool` function to support AI tool conversions, including automatic description generation based on resource and operation.
4. Added support for various data types (string, number, boolean, json, date) in AI-powered parameters.
5. Updated several nodes to be usable as AI tools.

### $fromAI Function

The `$fromAI()` function is implemented in the `WorkflowDataProxy` class and accepts four parameters:

1. `name`: A string identifier for the parameter (required)
2. `description`: A string describing the parameter's purpose (optional)
3. `type`: A string specifying the data type (optional, defaults to 'string')
4. `defaultValue`: A fallback value if no AI input is provided (optional)

Example usage:
```javascript
"inputText": "={{ $fromAI('inputText', 'Enter main text to process', 'string') }}"
```

### Parsing and Schema Generation

- The `parseFromAIArguments` function uses a regex to extract all `$fromAI()` calls from the node parameters.
- Each extracted call is parsed into a `FromAIArgument` object using the `parseArguments` function.
- The `generateZodSchema` function creates a Zod schema based on the parsed arguments, which is used for validation and type-checking.

### Node Conversion to AI Tools

The `convertNodeToAiTool` function has been updated to:

- Modify the node's name, displayName, and I/O configuration for AI tool usage.
- Add a `descriptionType` property allowing users to choose between automatic and manual description generation.
- Insert a `toolDescription` property for manual description input when needed.

### WorkflowDataProxy Updates

- The `$fromAI` method in `WorkflowDataProxy` retrieves values passed by LLMs from the `runExecutionData`.
- It includes validation for parameter names and handles cases where no execution data is available.

### CodeMirror Integration
CleanShot 2024-09-17 at 21 53 09

- Added `$fromAI()` to the list of root dollar completions in the CodeMirror plugin.
- Updated the completion info to include examples and parameter descriptions.

### Node Updates

Several core nodes (including Airtable, Baserow, Google Calendar, Gmail, Google Sheets, HackerNews, Postgres, Supabase, Telegram) have been updated with a `usableAsTool: true` property to indicate their compatibility with this new feature.

## Review / Merge checklist

- [x] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) 
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [x] Tests included. 
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

### CODE REVIEW

File: packages/core/src/CreateNodeAsTool.ts [hunk 1]
```diff
@@ -1,182 +1,244 @@
...
+	// This loop iterates through each character in the argsString
+	for (let i = 0; i < argsString.length; i++) {
+		const char = argsString[i];
 
-		// Build the nested structure
-		for (let i = 0; i < remainingPath.length - 1; i++) {
-			if (!(remainingPath[i] in current)) {
-				current[remainingPath[i]] = {};
+		// Handle quotes (single or double)
+		if (char === "'" || char === '"') {
```
Comment (jeanpaul) [inline packages/core/src/CreateNodeAsTool.ts:164]: Does this handle escaped characters properly? I tried it, and got some weird results:

```
argsString: 'database-filter', 'the filter\\'s keyword to apply to the database', string, ['a\\'b', 'b']
result: {
  key: 'database-filter',
  description: "the filter\\\\'s keyword to apply to the database', string, ['a\\\\'b",
  type: "b']",
  defaultValue: undefined
}
```

**Review by jeanpaul — CHANGES_REQUESTED** [review]
Bravo! This is awesome work!

I have a couple of smaller comments / questions, but one thing that we definitely need to address is the parsing of the arguments to `$fromAI`. One thing that came to mind is that we may need to do more proper parsing, because users might be inclined to do things we don't anticipate yet, that are valid javascript.

Also, as discussed, I encountered an issue with Claude, when returning plain results from the HN node; I'm not sure what to do there, but maybe we need to wrap the results in something that's more easily digestible for some models?

File: packages/workflow/src/NodeHelpers.ts [hunk 2]
```diff
@@ -383,7 +407,20 @@ export function convertNodeToAiTool<
 					'Explain to the LLM what this tool does, a good, specific description would allow LLMs to produce expected results much more often',
 				placeholder: `e.g. ${item.description.description}`,
 			};
-			item.description.properties.unshift(descProp);
+
+			item.description.properties.unshift(...[descProp]);
```
[the remark this answers is not in the document]
Comment (OlegIvaniv) [reply packages/workflow/src/NodeHelpers.ts]: It was a leftover from the older placeholder implementations, good catch!

File: packages/editor-ui/src/composables/useToast.ts [hunk 3]
```diff
@@ -164,7 +164,7 @@ export function useToast() {
 	}
 
 	function causedByCredential(message: string | undefined) {
-		if (!message) return false;
+		if (!message || typeof message !== 'string') return false;
```
[the remark this answers is not in the document]
Comment (OlegIvaniv) [reply packages/editor-ui/src/composables/useToast.ts:167]: boyscout-rule change :D

File: packages/core/src/CreateNodeAsTool.ts (hunk 1 above)
Comment (OlegIvaniv) [reply packages/core/src/CreateNodeAsTool.ts]: I improved the parsing method and added a few more tests; could you have another look?

**Comment by OlegIvaniv** [comment]
Thanks for the review @jeanpaul! I improved the parsing logic and added a few more tests, could you re-review, pls?

File: packages/core/src/CreateNodeAsTool.ts [hunk 4]
```diff
@@ -1,296 +1,432 @@
...
+				escapeNext = false;
+				continue;
+			}
+
+			if (char === '\\') {
+				escapeNext = true;
+				continue;
+			}
+
+			if (char === '"' || char === "'") {
```
Comment (jeanpaul) [inline packages/core/src/CreateNodeAsTool.ts:211]: should we add backticks as quote-chars as well?

**Review by jeanpaul — CHANGES_REQUESTED** [review]
Very nice! Awesome expansion of the tests as well.
Left another few small comments, but other than that looks great.

File: packages/core/src/CreateNodeAsTool.ts (hunk 4 above)
Comment (OlegIvaniv) [reply packages/core/src/CreateNodeAsTool.ts]: Updated + added unit tests

**Review by jeanpaul — CHANGES_REQUESTED** [review]
One small gotcha (and a quick check for my understanding :))

File: packages/core/src/CreateNodeAsTool.ts [hunk 5]
```diff
@@ -241,11 +233,13 @@ class AIParametersParser {
 			const trimmed = arg.trim();
 			if (
 				(trimmed.startsWith("'") && trimmed.endsWith("'")) ||
+				(trimmed.startsWith('`') && trimmed.endsWith('`')) ||
 				(trimmed.startsWith('"') && trimmed.endsWith('"'))
 			) {
 				return trimmed
 					.slice(1, -1)
 					.replace(/\\'/g, "'")
+					.replace(/\\"/g, '`')
```
[the remark this answers is not in the document]
Comment (OlegIvaniv) [reply packages/core/src/CreateNodeAsTool.ts]: Nice catch! Fixed

File: packages/core/src/CreateNodeAsTool.ts [hunk 6]
```diff
@@ -326,14 +320,42 @@ class AIParametersParser {
...
 			}
+
+			if (keyMap.has(argument.key)) {
+				// If the key already exists in the Map
+				const existingArg = keyMap.get(argument.key)!;
+
+				// Check if the existing argument has the same description and type
+				if (
+					existingArg.description !== argument.description ||
+					existingArg.type !== argument.type
```
[the remark this answers is not in the document]
Comment (OlegIvaniv) [reply packages/core/src/CreateNodeAsTool.ts:342]: Yy, that's right. The final value is resolved in WorkflowDataProxy's `handleFromAi`

**Review by jeanpaul — APPROVED** [review]

**Comment by janober** [comment]
Got released with [`n8n@1.62.1`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.62.1)

**Pull request merged**